Flaky: SDK conformance tests

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via 
-->

`make run-sdk-conformance-test-go` is failing pretty often on CI builds.

**What happened**:

One of Sidecar methods (`Ready` in the logs below) call was not counted to the result.

**What you expected to happen**:

No errors on SDK tests. Add readability to test failure output.

**How to reproduce it (as minimally and precisely as possible)**:
`make run-sdk-conformance-test-go`

**Anything else we need to know?**:

Logs from https://console.cloud.google.com/cloud-build/builds/309432a8-a2c6-4f45-95f3-8801af9f04da;step=23?project=agones-images
```
includes/sdk.mk:162: recipe for target 'run-sdk-conformance-test-go' failed
includes/sdk.mk:152: recipe for target 'run-sdk-conformance-test' failed
make[1]: *** [run-sdk-conformance-test] Error 2
includes/sdk.mk:145: recipe for target 'run-sdk-conformance-no-build' failed
make[2]: *** [run-sdk-conformance-no-build] Error 1
{"message":"Testing Failed [ready allocate setlabel setannotation gameserver health shutdown watch reserve] [watch reserve allocate gameserver health setlabel setannotation shutdown]","severity":"info","time":"2020-04-06T19:33:00.434023752Z"}
{"message":"Compare","severity":"info","time":"2020-04-06T19:33:00.433934282Z"}
{"message":"shutting down sdk server","severity":"info","source":"main","time":"2020-04-06T19:33:00.433799498Z"}
```


**Environment**:
- Agones version: 
- Kubernetes version (use `kubectl version`): 1.14
- Cloud provider or hardware configuration: 
- Install method (yaml/helm): CI
- Troubleshooting guide log(s):
- Others:

--- a/pkg/sdkserver/localsdk.go
+++ b/pkg/sdkserver/localsdk.go
@@ -75,6 +75,7 @@ type LocalSDKServer struct {
 	gs                *sdk.GameServer
 	update            chan struct{}
 	updateObservers   sync.Map
+	testMutex         sync.Mutex
 	requestSequence   []string
 	expectedSequence  []string
 	gsState           agonesv1.GameServerState
@@ -90,6 +91,7 @@ func NewLocalSDKServer(filePath string) (*LocalSDKServer, error) {
 		gs:              defaultGs,
 		update:          make(chan struct{}),
 		updateObservers: sync.Map{},
+		testMutex:       sync.Mutex{},
 		requestSequence: make([]string, 0),
 		testMode:        false,
 		gsState:         agonesv1.GameServerStateScheduled,
@@ -164,6 +166,8 @@ func (l *LocalSDKServer) SetExpectedSequence(sequence []string) {
 // recordRequest append request name to slice
 func (l *LocalSDKServer) recordRequest(request string) {
 	if l.testMode {
+		l.testMutex.Lock()
+		defer l.testMutex.Unlock()
 		l.requestSequence = append(l.requestSequence, request)
 	}
 }
@@ -185,6 +189,8 @@ func (l *LocalSDKServer) recordRequestWithValue(request string, value string, ob
 		}
 
 		if value == fieldVal {
+			l.testMutex.Lock()
+			defer l.testMutex.Unlock()
 			l.requestSequence = append(l.requestSequence, request)
 		} else {
 			fmt.Printf("Error: we expected to receive '%s' as value for '%s' request but received '%s'. \n", fieldVal, request, value)

--- a/pkg/sdkserver/localsdk_test.go
+++ b/pkg/sdkserver/localsdk_test.go
@@ -17,6 +17,7 @@ package sdkserver
 import (
 	"context"
 	"encoding/json"
+	"fmt"
 	"io/ioutil"
 	"os"
 	"sync"
@@ -379,4 +380,40 @@ func TestLocalSDKServerStateUpdates(t *testing.T) {
 	gs, err = l.GetGameServer(ctx, e)
 	assert.Nil(t, err)
 	assert.Equal(t, gs.Status.State, string(agonesv1.GameServerStateShutdown))
+}
+
+// TestSDKConformanceFunctionality - run a number of record requests in parallel
+func TestSDKConformanceFunctionality(t *testing.T) {
+	t.Parallel()
+
+	l, err := NewLocalSDKServer("")
+	assert.Nil(t, err)
+	l.testMode = true
+	l.recordRequest("")
+	l.gs = &sdk.GameServer{ObjectMeta: &sdk.GameServer_ObjectMeta{Name: "empty"}}
+	exampleUID := "052fb0f4-3d50-11e5-b066-42010af0d7b6"
+	// field which is tested
+	setAnnotation := "setannotation"
+	l.gs.ObjectMeta.Uid = exampleUID
+
+	expected := []string{}
+	expected = append(expected, "", setAnnotation)
+
+	wg := sync.WaitGroup{}
+	for i := 0; i < 20; i++ {
+		wg.Add(1)
+		str := fmt.Sprintf("%d", i)
+		expected = append(expected, str)
+
+		go func() {
+			l.recordRequest(str)
+			l.recordRequestWithValue(setAnnotation, exampleUID, "UID")
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	l.SetExpectedSequence(expected)
+	b := EqualSets(l.expectedSequence, l.requestSequence)
+	assert.True(t, b, "we should receive strings from all go routines %v %v", l.expectedSequence, l.requestSequence)
 }

